### PR TITLE
Issue #15429: Added SuppressionXpathSingleFilter suppression ignore Test Methods for section 5.2.3 Method Names

### PR DESCRIPTION
**Pull request by Zopsss**
#15429 

Added [SuppressionXpathSingleFilter](https://checkstyle.sourceforge.io/filters/suppressionxpathsinglefilter.html) suppression in google_checks.xml and updated regex at MethodName module to allow test methods to have `_` in their name for section [5.2.3 Method Names](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/bcdd4d1_2024123825/google_style.html#a5.2.3), for other methods, module will give violation if `_` is used in method name

### CODE REVIEW

File: src/it/resources/com/google/checkstyle/test/chapter5naming/rule523methodnames/InputMethodName.java [hunk 1]
```diff
@@ -42,6 +42,9 @@ void testingFoo_foo() {}
   @Test
   void testing_0123() {}
 
+  @Test
+  void Testing_Foo_False_Negative() {} // false negative, should be a violation
```
Comment (Zopsss) [inline src/it/resources/com/google/checkstyle/test/chapter5naming/rule523methodnames/InputMethodName.java:46]: This is a false negative. If there's `_` in a test method name then we're completely ignoring the method. Is there a workaround for this? @romani

File: src/it/resources/com/google/checkstyle/test/chapter5naming/rule523methodnames/InputMethodName.java [hunk 2]
```diff
@@ -42,6 +42,11 @@ void testingFoo_foo() {}
...
+
+  void Testing_Foo() {} // violation 'Method name 'Testing_Foo' must match pattern'
```
Comment (Zopsss) [inline src/it/resources/com/google/checkstyle/test/chapter5naming/rule523methodnames/InputMethodName.java:48]: Though we're able to detect this for normal methods

**Review by Zopsss — COMMENTED** [review]
False Negative:

File: src/main/resources/google_checks.xml [hunk 3]
```diff
@@ -105,6 +105,12 @@
       <property name="query" value="//RCURLY[parent::SLIST[count(./*)=1]
                                      or preceding-sibling::*[last()][self::LCURLY]]"/>
     </module>
+    <module name="SuppressionXpathSingleFilter">
+      <property name="checks" value="MethodName"/>
+      <property name="query" value="//METHOD_DEF[
+                                     .//ANNOTATION/IDENT[@text='Test']]//*"/>
+      <property name="message" value=".*_.*"/>
```
Comment (romani) [inline src/main/resources/google_checks.xml:112]: Let's make message match to be
`.*'[a-z][a-z0-9][_a-zA-Z0-9]*'.*`
It should not suppress your case `@Test  void Testing_Foo_False_Negative()`

**Review by romani — CHANGES_REQUESTED** [review]
Good.

Item:

File: src/it/resources/com/google/checkstyle/test/chapter5naming/rule523methodnames/InputMethodName.java (hunk 1 above)
Comment (romani) [reply src/it/resources/com/google/checkstyle/test/chapter5naming/rule523methodnames/InputMethodName.java:46]: See below.
https://github.com/checkstyle/checkstyle/pull/15515#discussion_r1720518410

File: src/main/resources/google_checks.xml [hunk 4]
```diff
@@ -105,6 +105,12 @@
       <property name="query" value="//RCURLY[parent::SLIST[count(./*)=1]
                                      or preceding-sibling::*[last()][self::LCURLY]]"/>
     </module>
+    <module name="SuppressionXpathSingleFilter">
+      <property name="checks" value="MethodName"/>
+      <property name="query" value="//METHOD_DEF[
+                                     .//ANNOTATION/IDENT[@text='Test']]//*"/>
+      <property name="message" value="'[a-z][a-z0-9][_a-zA-Z0-9]*'.*"/>
```
Comment (Zopsss) [inline src/main/resources/google_checks.xml:112]: removed the preceding `.*` from suggested regex:

```
.*'[a-z][a-z0-9][_a-zA-Z0-9]*'.*
```

I don't think it's important

File: src/it/resources/com/google/checkstyle/test/chapter5naming/rule523methodnames/InputMethodName.java [hunk 5]
```diff
@@ -42,6 +42,17 @@ void testingFoo_foo() {}
...
   void testing_0123() {}
 
+  @Test
+  void Testing_Foo() {} // violation 'Method name 'Testing_Foo' must match pattern'
+
+  @Test
+  void t_esting() {} // violation 'Method name 't_esting' must match pattern'
+
+  @Test
+  void _testing() {} // violation 'Method name '_testing' must match pattern'
```
Comment (Zopsss) [inline src/it/resources/com/google/checkstyle/test/chapter5naming/rule523methodnames/InputMethodName.java:52]: Now we're able to detect `Testing_Foo`. We're giving violation for `t_esting` and `_testing`, there's no explicit rule specified for placement of `_` in test method name, the initial character of test method is also lowercase character so ideally we should not give violation for these methods but no one writes names like this and no one should actually do, I think we should keep this violation but we should add blue tick for MethodName module on coverage page

File: src/main/resources/google_checks.xml (hunk 4 above)
Comment (Zopsss) [inline src/main/resources/google_checks.xml:112]: removed the preceding `.*` from suggested regex:

```
.*'[a-z][a-z0-9][_a-zA-Z0-9]*'.*
```

I don't think it's important

**Review by Zopsss — COMMENTED** [review]
New changes:

File: src/main/resources/google_checks.xml (hunk 3 above)
Comment (Zopsss) [reply src/main/resources/google_checks.xml]: done, changed regex a little bit, explained here: https://github.com/checkstyle/checkstyle/pull/15515#discussion_r1720790213

File: src/it/resources/com/google/checkstyle/test/chapter5naming/rule523methodnames/InputMethodName.java (hunk 5 above)
Comment (romani) [reply src/it/resources/com/google/checkstyle/test/chapter5naming/rule523methodnames/InputMethodName.java:52]: yes, such crazy names are ok to violate.

File: src/main/resources/google_checks.xml (hunk 4 above)
Comment (romani) [reply src/main/resources/google_checks.xml:112]: yes

**Comment by romani** [comment]
CI failure is unrelated.

**Review by romani — APPROVED** [review]
ok to merge

**Review by rdiachenko — APPROVED** [review]
lgtm

**Pull request merged**